src/hazelcore/Builtins.re

```diff
@@ -0,0 +1,181 @@
+type form = (
+  /* eval: */ (args, DHExp.t => EvaluatorResult.t) => EvaluatorResult.t,
+  /* elab: */ DHExp.t,
+);
+
+[@deriving sexp]
+type t = (Var.t, HTyp.t, form);
+
+module Impl = {
+  [@deriving sexp]
+  type f =
+    (/* args: */ args, /* evaluate: */ DHExp.t => EvaluatorResult.t) =>
+    EvaluatorResult.t;
+
+  /*
+     Build the elaborated DHExp for a built-in function.
+
+     For example:
+       mk_elab("mod", Arrow(Int, Arrow(Int, Int)))
+         =>
+          Lam("x0", Arrow(Int, Arrow(Int, Int)),
+            Lam("x1", Arrow(Int, Int),
+              Apbuilt-in("mod", [BoundVar("x0"), BoundVar("x1")])
+            )
+          )
+   */
+  let mk_elab = (ident: Var.t, ty: HTyp.t): DHExp.t => {
+    let rec mk_elab_inner =
+            (ty': HTyp.t, n: int, bindings: list(Var.t)): DHExp.t => {
+      switch (ty') {
+      | Arrow(_, ty'') =>
+        let var = "x" ++ string_of_int(n);
+        Lam(Var(var), ty', mk_elab_inner(ty'', n + 1, [var, ...bindings]));
+      | _ =>
+        let bindings = List.rev_map(x => DHExp.BoundVar(x), bindings);
+        ApBuiltin(ident, bindings);
+      };
+    };
+
+    mk_elab_inner(ty, 0, []);
+  };
+
+  /*
+     Create a built-in function.
+   */
+  let mk = (ident: Var.t, ty: HTyp.t, fn: f): t => {
+    let eval = (args, evaluate) => fn(args, evaluate);
+    let elab = mk_elab(ident, ty);
+
+    (ident, ty, (eval, elab));
+  };
+
+  /*
+     Create a built-in constant.
+   */
+  let mk_zero = (ident: Var.t, ty: HTyp.t, v: DHExp.t): t => {
+    let fn = (args, evaluate) => {
+      switch (args) {
+      | [] => evaluate(v)
+      | _ => raise(EvaluatorError.Exception(BadBuiltinAp(ident, args)))
+      };
+    };
+
+    mk(ident, ty, fn);
+  };
+
+  /*
+     Create a built-in function that takes a single argument. The given type
+     must be correct.
+   */
+  let mk_one =
+      (
+        ident: Var.t,
+        ty: HTyp.t,
+        fn: (Var.t, EvaluatorResult.t) => EvaluatorResult.t,
+      )
+      : t => {
+    let fn = (args, evaluate) => {
+      switch (args) {
+      | [d1] =>
+        let d1' = evaluate(d1);
+        fn(ident, d1');
+      | _ => raise(EvaluatorError.Exception(BadBuiltinAp(ident, args)))
+      };
+    };
+
+    mk(ident, ty, fn);
+  };
+
+  /*
+     Create a built-in function that takes two arguments. The given type must be
+     correct.
+   */
+  let mk_two =
+      (
+        ident: Var.t,
+        ty: HTyp.t,
+        fn: (Var.t, EvaluatorResult.t, EvaluatorResult.t) => EvaluatorResult.t,
+      )
+      : t => {
+    let fn = (args, evaluate) => {
+      switch (args) {
+      | [d1, d2] =>
+        let d1' = evaluate(d1);
+        let d2' = evaluate(d2);
+        fn(ident, d1', d2');
+      | _ => raise(EvaluatorError.Exception(BadBuiltinAp(ident, args)))
+      };
+    };
+
+    mk(ident, ty, fn);
+  };
+};
+
+module Impls = {
+  open EvaluatorResult;
+
+  /* int_of_float implementation. */
+  let int_of_float = (ident, d1') =>
+    switch (d1') {
```

Comment:
Let's use `r` variables for evaluator results and keep `d` only for DHExps so that its easier to read